### PR TITLE
150 remove config

### PR DESCRIPTION
**Pull request by RobinFeldmann**
TLDR:
- Remove the HW-NAS and DeploymentConfig
- Replace by DockerParams, SSHParams and SerialParams, HardwareConstraints, HWNASParameters Dataclasses to structure the parameterization of HW-NAS search and Deployment
- Remove all yaml-file in tests
  - Add a simple system_test_settings.py for setting parameters for connecting to PICO and RPI Device.
- Fixing issues with optuna Gridsearch and the explorer searchspace.
  -  adding SearchSpace.to_grid() ->  dict[str, Sequence[Any]] to create a grid of all parameter choices
- #155
  - make constraints (hard and soft) completely modular by adding a ConstraintRegistry where you can add constraints to specific Estimators. These are automatically evaluated in the hw_nas.search().

### CODE REVIEW

File: elasticai/explorer/hw_nas/search_space/construct_search_space.py [hunk 1]
```diff
@@ -35,6 +36,57 @@ def createLinear(self, trial, block, num_layers, search_params):
...
             self.layers.append(activation_mapping[activation])
 
+    def to_grid(self) -> dict[str, Sequence[Any]]:
+        def parse_params_for_grid(param) -> list[Any]:
+            # Expand ranges into an list of possible values
+            if isinstance(param, list):
+                return param
+            elif isinstance(param, dict) and "start" in param and "end" in param:
+                if isinstance(param["start"], int):
+                    return list(range(param["start"], param["end"] + 1))
```
Comment (chrisringhofer) [inline elasticai/explorer/hw_nas/search_space/construct_search_space.py]: Double-check and consolidate with parse_search_params

File: elasticai/explorer/hw_nas/hw_nas.py [hunk 2]
```diff
@@ -28,75 +30,65 @@ class SearchAlgorithm(Enum):
...
+
+            soft_constraints = constraint_registry.get_soft_constraints(estimator)
+            for sc in soft_constraints:
+                raw_value = sc.penalty_fn(
+                    sc.estimate_transform(estimate), sc.constraint_value
+                )
+
+                # a reward is handled as a negative penalty
+                penalty_amount = -raw_value if sc.is_reward else raw_value
+                soft_score -= penalty_amount * sc.weight
```
Comment (chrisringhofer) [inline elasticai/explorer/hw_nas/hw_nas.py]: Das hier ändert sich vermutlich, wenn wir "Objectives" als Konzept einführen bzw. von "Soft Constraints" trennen

File: elasticai/explorer/hw_nas/estimators.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+from torch.optim.adam import Adam
+from elasticai.explorer.training.data import DatasetSpecification
+from elasticai.explorer.training.trainer import Trainer
+
+handlers = {"aten::sigmoid": None}
+
+
+class Estimator:
+    def __init__(self) -> None:
+        self.metric_id: str = "estimate"
```
Comment (chrisringhofer) [inline elasticai/explorer/hw_nas/estimators.py]: Perhaps rename in "metric_name"

File: elasticai/explorer/hw_nas/hw_nas.py [hunk 4]
```diff
@@ -150,17 +143,18 @@ def search(
...
     for frozen_trial in top_k_frozen_trials:
         top_k_models.append(search_space.create_model_sample(frozen_trial))
         top_k_params.append(frozen_trial.params)
         top_k_metrics.append(
             {
-                "default": eval(frozen_trial),
-                "val_accuracy": frozen_trial.user_attrs["val_accuracy"],
-                "val_loss": frozen_trial.user_attrs["val_loss"],
-                "flops log10": frozen_trial.user_attrs["flops_log10"],
+                "soft_score": eval(frozen_trial),
```
Comment (chrisringhofer) [inline elasticai/explorer/hw_nas/hw_nas.py]: Wird vermutlich dann einfach "score" (also objective(s) minus penalty)

File: README.md [hunk 5]
```diff
@@ -39,4 +39,4 @@ If you don't need dev dependencies add:
 
 # Examples 
 For the full workflow from HW-NAS to Tests on Device, see the "examples" Folder.
-To configure the example for your specific setup, create your own "hwnas_config.yaml" and "deployment_config.yaml" with the help of the templates in "config_files/config_templates".
\ No newline at end of file
+To configure the example for your specific setup, create your own "deployment_config.yaml" with the help of the templates in "config_files/config_templates".
```
Comment (chrisringhofer) [inline README.md]: deployment_config auch raus :)

**Review by chrisringhofer — CHANGES_REQUESTED** [review]
Comments bearbeiten :) Sonst echt cool, danke!

File: elasticai/explorer/hw_nas/hw_nas.py [hunk 6]
```diff
@@ -1,111 +1,106 @@
...
-    constraints: dict[str, int] = {},
+    optimization_criteria_registry: OptimizationCriteriaRegistry,
 ) -> float:
 
     def objective(trial: optuna.Trial) -> float:
 
         search_space = SearchSpace(search_space_cfg)
-        try:
-            model = search_space.create_model_sample(trial)
-        except NotImplementedError:
```
Comment (mokouMonday) [inline elasticai/explorer/hw_nas/hw_nas.py]: Das muss da rein, sonst bricht die Suche ab, sobald eine Kombi nicht implementiert ist

File: tests/integration_tests/test_hw_nas_setup_and_search.py [hunk 7]
```diff
@@ -58,32 +65,59 @@ def setup_class(self):
...
+            (SearchStrategy.EVOlUTIONARY_SEARCH, False, 2),
+            (SearchStrategy.RANDOM_SEARCH, True, 0),
         ],
     )
-    def test_search(self, hwnas_cfg, expected):
-
-        hwnas_cfg = HWNASConfig(config_path=ROOT_DIR / Path(hwnas_cfg))
-        self.RPI5explorer.generate_search_space(
-            Path(ROOT_DIR / "elasticai/explorer/hw_nas/search_space/search_space.yaml")
+    def test_search(self, search_strategy, with_hardconstraints, expected):
```
[the remark this answers is not in the document]
Comment (RobinFeldmann) [reply tests/integration_tests/test_hw_nas_setup_and_search.py]: Also der testet ob die Hardconstraints funktionieren, und die Suchstrategien mit Accuracy Estimation. Das sind die Tests die Chris geschrieben hat, mit @pytest.mark.parametrize umgeschrieben und aufs wesentliche reduziert.

File: elasticai/explorer/hw_nas/estimators.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    ) -> tuple[float | int, list[float | int]]:
+        param_count = parameter_count(model_sample)[""]
+        return param_count, []
+
+
+class TrainMetricsEstimator(Estimator):
+    def __init__(
+        self,
+        trainer: Trainer,
+        metric_name: str | Literal["loss"] = "loss",
```
Comment (chrisringhofer) [inline elasticai/explorer/hw_nas/estimators.py]: Literal wech

File: elasticai/explorer/hw_nas/hw_nas.py [hunk 9]
```diff
@@ -1,111 +1,106 @@
...
+                    )
+
+            objectives = optimization_criteria_registry.get_objectives(estimator)
+            for o in objectives:
+                if o.transform:
+                    objective_value = o.weight * o.transform(estimate)
+                else:
+                    objective_value = o.weight * estimate
+
+                score += objective_value
```
[the remark this answers is not in the document]
Comment (chrisringhofer) [reply elasticai/explorer/hw_nas/hw_nas.py]: Resolved by creation of issue #162
